### DISCUSSION TITLE
Valent Status (October 2024)

### DISCUSSION THREAD

**Discussion by andyholmes** [andyholmes/valent#163] [Announcements]
## 👋 Welcome to the Valent project!

Valent is a very functional [KDE Connect](https://kdeconnect.kde.org) client, with a pretty good [test suite](https://valent.andyholmes.ca/coverage), but is still under heavy development. [Issues](https://github.com/andyholmes/valent/issues) are accepted, but there are no guarantees for when they may be addressed.

Feature requests, suggestions and support requests should be posted to here in the [Discussions](https://github.com/andyholmes/valent/discussions) board. Issues are intended for reporting problems and anything misfiled there will be moved to a new discussion.

The project is also open for [contributions](https://github.com/andyholmes/valent/blob/main/CONTRIBUTING.md), but let me know in case I have something planned or half-finished.

### Tips

1. **Expect Bugs and Missing Features**

   There are aspects of Valent that are intentionally broken, either temporarily or as an expected result of development. The cache and databases are very unstable, and will continue to be broken with no migration planned while the project is in alpha.

2. **Clear Your Cache**

   When something goes wrong, try clearing the the cache. Your **devices will not be unpaired** and any cached data will re-sync as necessary when the device is connected.

   First, ensure that Valent isn't running by using the *Quit* menu item, or running `killall -15 valent`. When you restart Valent, the device list will not display your paired devices until they're discovered on the network.

   |         |                                                             |
   | ------- | ----------------------------------------------------------- |
   | GNOME   | Clear the cache via the **Apps** page in **GNOME Settings** |
   | Flatpak | `rm -rf ~/.var/app/ca.andyholmes.Valent/cache/valent`       |
   | Other   | `rm -rf ~/.cache/valent`                                    |

3. **Open "Atomic Issues"**

   [Atomic commits] is the practice of grouping only related changes into a commit, which makes tracking progress and regressions much easier. So, the best presented issues can be closed by a commit titled `fix: $ISSUE_TITLE`.

   If the problem you're reporting doesn't seem "atomic", break it into separate issues or start a discussion about the larger problem and we can sort it out together.

[Atomic commits]: https://en.wikipedia.org/wiki/Atomic_commit#Atomic_commit_convention

#### Q: Does Valent replace GSConnect?

If you are looking for a stable KDE Connect client, I recommend [KDE Connect](https://kdeconnect.kde.org). 

Valent has a newer code base and better tests than [GSConnect](https://github.com/GSConnect/gnome-shell-extension-gsconnect), but very little real world testing and no commitment to stability.

#### Q: Does Valent support `X`?

Valent is a KDE Connect client and only supports protocol features available in the reference implementation (i.e. `kdeconnect-android` and `kdeconnect-kde`). It's possible for these features to be implemented as third-party plugins, but Valent will never include features that break or extend the KDE Connect protocol.

Any new features can be discussed upstream with the [KDE Connect team](https://kdeconnect.kde.org/get-involved.html), where GNOME, KDE, Android, iOS/macOS, Sailfish and Windows developers can all collaborate on a protocol that works everywhere.

#### Q: Valent for non-GNOME environments?

The core of Valent is written in C with the GNOME platform libraries and does not (necessarily) depend on a GUI toolkit. Most functionality is implemented in plugins, including the `gnome` plugin which provides the default GTK4/Adwaita UI.

An alternate UI plugin (plain GTK4, Granite, TUI, etc) **may** be accepted into the main project, if it is well written and comes with a reasonable commitment to maintain it. If you have plans to contribute such a plugin, please reach out by email or in discussions.

### What's the Plan? What's Next? How long?

As with most projects, there are no ETAs. Since there are several KDE Connect clients available already, there is no rush to release a stable version of Valent.

Hard blockers for a beta release include:

* Accessibility

   Acceptable support for AT devices (e.g. screen readers) is a requirement for a beta release, so that it is usable enough for those with AT devices to test, report issues and comment on the general UX.
* User Interface

   Before the accessibility can be reviewed, the default user interface needs to be mostly complete. Some elements like messaging and remote control involve fairly complex UI and may take some time.
* API Stability

   Depending on the amount of interest in third-party plugins, a beta release may be delayed until the API is stable. If there is little or no interest, the plugin API may remain somewhat unstable or just kept private.

**Comment by andyholmes** [comment]
Here's a few screenshots, for those that are into those kind of things. This doesn't cover everything, but should give a decent idea of how the application looks in practice.

## Main Page / Application Preferences
![1659904299_grim](https://user-images.githubusercontent.com/1265208/183309999-43700a00-ce9a-4aef-b675-524c5d2c38e8.png)![1659903608_grim](https://user-images.githubusercontent.com/1265208/183309884-0d11dd35-29f9-483e-86b2-3554dfe9a739.png)

## Device Page (with Connectivity Report) / Presentation Remote
![1659903633_grim](https://user-images.githubusercontent.com/1265208/183309886-c5575d0b-dd16-47f9-aa0b-4dc723bef47c.png)![1659903683_grim](https://user-images.githubusercontent.com/1265208/183309892-c3ecae6a-8427-4fae-a1ad-8a465ebe6e16.png)

## Device Preferences / Battery Plugin Preferences
![1659903643_grim](https://user-images.githubusercontent.com/1265208/183309889-5aeceebe-1ba9-42d1-9287-0e4712f50951.png)![1659903665_grim](https://user-images.githubusercontent.com/1265208/183309891-23536b3c-6c5f-4db4-aa9e-8143bf486017.png)

**Comment by andyholmes** [comment]
New feature! :slightly_smiling_face: 

I've added a media remote, like the one found in KDE Connect Android. It seems to work pretty well, although there are still a few impedance issues to work out.

Most (all?) MPRIS players seem to be uncompliant in at least one way, on top of the fact that Android media players are not MPRIS players but mapped to MPRIS functionality via KDE Connect's JSON protocol. KDE Connect Android also doesn't seem to support a few things like uploading album art or receiving relative "seek" commands.

If this is a useful feature to you, feel free to start a dedicated discussion with mockups, strange behaviour or suggestions :slightly_smiling_face: 

![Screenshot from 2022-12-19 16-23-12](https://user-images.githubusercontent.com/1265208/208553201-40cf3e7d-6eb2-4af9-9015-f2bbc3b99d68.png)

**Reply by 000exploit** [reply]
Can it create multiple MPRIS players, or at least switch them more often so, for example, when Spotify is playing on another device, don't have control over NewPipe?

**Reply by andyholmes** [reply]
Multiple players should be exported if not running in a Flatpak, otherwise it should export the player that has most recently been in a "Play" state.

**Reply by 000exploit** [reply]
Is this flatpak limitation?

**Reply by andyholmes** [reply]
Flatpak doesn't prevent Valent from requesting the permission, if that's what you mean.
